**Employee Data Cleaning & Analysis using Pandas**

Subtitle:
A beginner-friendly data cleaning and exploratory analysis project using Python Pandas.

## Project Overview

This project focuses on cleaning and analyzing an employee dataset using Python and Pandas.

### Objectives
- Handle missing values
- Clean inconsistent text data
- Check data types
- Identify duplicates
- Sort and filter data
- Perform basic salary and experience analysis
- Extract meaningful insights from the dataset

## Tools & Technologies

- Python
- Pandas
- Google Colab

## Dataset

The dataset contains employee information such as:
- Employee Name
- Salary
- Department
- Experience
- City

In [3]:
import pandas as pd

data = {
    "Employee_Name": [" Rohit Sharma", "virat kohli ", "RAHUL", " Rohit Sharma", "shubman gill"],
    "Salary": [25000, 45000, None, 25000, 60000],
    "Department": ["IT", "it", "HR ", "IT", None],
    "Experience": [2, 5, 3, 2, None],
    "City": [" Jaipur ", "Delhi", " jaipur", "Mumbai", "DELHI"]
}

df = pd.DataFrame(data)

df

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,IT,2.0,Jaipur
1,virat kohli,45000.0,it,5.0,Delhi
2,RAHUL,NaN,HR,3.0,jaipur
3,Rohit Sharma,25000.0,IT,2.0,Mumbai
4,shubman gill,60000.0,None,NaN,DELHI


## Data Cleaning

In [4]:
df.isna().sum()

,0
Employee_Name,0
Salary,1
Department,1
Experience,1
City,0


In [5]:
print(df["Salary"].mean())

38750.0


In [6]:
df["Salary"] = df["Salary"].fillna(df["Salary"].mean())

In [7]:
df

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,IT,2.0,Jaipur
1,virat kohli,45000.0,it,5.0,Delhi
2,RAHUL,38750.0,HR,3.0,jaipur
3,Rohit Sharma,25000.0,IT,2.0,Mumbai
4,shubman gill,60000.0,None,NaN,DELHI


In [8]:
df["Department"] = df["Department"].fillna("IT")

In [9]:
df

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,IT,2.0,Jaipur
1,virat kohli,45000.0,it,5.0,Delhi
2,RAHUL,38750.0,HR,3.0,jaipur
3,Rohit Sharma,25000.0,IT,2.0,Mumbai
4,shubman gill,60000.0,IT,NaN,DELHI


In [10]:
df["Experience"].mean()

np.float64(3.0)

In [11]:
df["Experience"] = df["Experience"].fillna(df["Experience"].mean())

In [12]:
df

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,IT,2.0,Jaipur
1,virat kohli,45000.0,it,5.0,Delhi
2,RAHUL,38750.0,HR,3.0,jaipur
3,Rohit Sharma,25000.0,IT,2.0,Mumbai
4,shubman gill,60000.0,IT,3.0,DELHI


In [13]:
df.isna().sum()

,0
Employee_Name,0
Salary,0
Department,0
Experience,0
City,0


In [14]:
df["Employee_Name"] = df['Employee_Name'].str.strip()

In [15]:

df

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,IT,2.0,Jaipur
1,virat kohli,45000.0,it,5.0,Delhi
2,RAHUL,38750.0,HR,3.0,jaipur
3,Rohit Sharma,25000.0,IT,2.0,Mumbai
4,shubman gill,60000.0,IT,3.0,DELHI


In [16]:
df["Employee_Name"].iloc[0]

'Rohit Sharma'

In [17]:
df["City"] = df["City"].str.strip()

In [18]:
df

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,IT,2.0,Jaipur
1,virat kohli,45000.0,it,5.0,Delhi
2,RAHUL,38750.0,HR,3.0,jaipur
3,Rohit Sharma,25000.0,IT,2.0,Mumbai
4,shubman gill,60000.0,IT,3.0,DELHI


In [19]:
df["Department"] = df["Department"].str.strip()

In [20]:

df


,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,IT,2.0,Jaipur
1,virat kohli,45000.0,it,5.0,Delhi
2,RAHUL,38750.0,HR,3.0,jaipur
3,Rohit Sharma,25000.0,IT,2.0,Mumbai
4,shubman gill,60000.0,IT,3.0,DELHI


In [21]:
df.columns

Index(['Employee_Name', 'Salary', 'Department', 'Experience', 'City'], dtype='object')

In [22]:

df

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,IT,2.0,Jaipur
1,virat kohli,45000.0,it,5.0,Delhi
2,RAHUL,38750.0,HR,3.0,jaipur
3,Rohit Sharma,25000.0,IT,2.0,Mumbai
4,shubman gill,60000.0,IT,3.0,DELHI


In [23]:
df["Employee_Name"] = df["Employee_Name"].str.title()

In [24]:
df["Department"] = df["Department"].str.title()

In [25]:
df["City"] = df["City"].str.title()

In [26]:
df["Salary"].dtype


dtype('float64')

In [27]:
df["Salary"]

,Salary
0,25000.0
1,45000.0
2,38750.0
3,25000.0
4,60000.0


In [28]:
df.duplicated()

,0
0,False
1,False
2,False
3,False
4,False


In [29]:
df.duplicated(subset=["Employee_Name"])

,0
0,False
1,False
2,False
3,True
4,False


In [30]:
df


,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,It,2.0,Jaipur
1,Virat Kohli,45000.0,It,5.0,Delhi
2,Rahul,38750.0,Hr,3.0,Jaipur
3,Rohit Sharma,25000.0,It,2.0,Mumbai
4,Shubman Gill,60000.0,It,3.0,Delhi


In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Employee_Name  5 non-null      object 
 1   Salary         5 non-null      float64
 2   Department     5 non-null      object 
 3   Experience     5 non-null      float64
 4   City           5 non-null      object 
dtypes: float64(2), object(3)
memory usage: 332.0+ bytes


In [32]:
df.sort_values("Salary")

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,It,2.0,Jaipur
3,Rohit Sharma,25000.0,It,2.0,Mumbai
2,Rahul,38750.0,Hr,3.0,Jaipur
1,Virat Kohli,45000.0,It,5.0,Delhi
4,Shubman Gill,60000.0,It,3.0,Delhi


## Data Analysis

In [34]:
df["Department"].value_counts()

,count
Department,
It,4
Hr,1


In [35]:
df["Department"] = df["Department"].str.title()

In [36]:
df

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,It,2.0,Jaipur
1,Virat Kohli,45000.0,It,5.0,Delhi
2,Rahul,38750.0,Hr,3.0,Jaipur
3,Rohit Sharma,25000.0,It,2.0,Mumbai
4,Shubman Gill,60000.0,It,3.0,Delhi


In [37]:
df["Department"].value_counts()

,count
Department,
It,4
Hr,1


In [38]:
df["Salary"].max()

60000.0

In [39]:
df.loc[df["Salary"].idxmax()]

,4
Employee_Name,Shubman Gill
Salary,60000.0
Department,It
Experience,3.0
City,Delhi


In [40]:
df

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,It,2.0,Jaipur
1,Virat Kohli,45000.0,It,5.0,Delhi
2,Rahul,38750.0,Hr,3.0,Jaipur
3,Rohit Sharma,25000.0,It,2.0,Mumbai
4,Shubman Gill,60000.0,It,3.0,Delhi


In [41]:
df["Salary"].mean()

np.float64(38750.0)

In [42]:
df["Salary"].max() - df["Salary"].min()

35000.0

In [43]:
df["Salary"].unique()


array([25000., 45000., 38750., 60000.])

In [44]:
df["Salary"].nunique()

4

In [45]:
df["Salary"].value_counts()

,count
Salary,
25000.0,2
45000.0,1
38750.0,1
60000.0,1


In [46]:
df

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,It,2.0,Jaipur
1,Virat Kohli,45000.0,It,5.0,Delhi
2,Rahul,38750.0,Hr,3.0,Jaipur
3,Rohit Sharma,25000.0,It,2.0,Mumbai
4,Shubman Gill,60000.0,It,3.0,Delhi


In [47]:
df['City'].value_counts()

,count
City,
Jaipur,2
Delhi,2
Mumbai,1


In [48]:
df["Employee_Name"].size

5

In [49]:
df["Experience"].mean()

np.float64(3.0)

In [50]:
df.loc[df["Experience"].idxmax()]

,1
Employee_Name,Virat Kohli
Salary,45000.0
Department,It
Experience,5.0
City,Delhi


In [51]:
df.loc[df["Experience"].idxmin()]

,0
Employee_Name,Rohit Sharma
Salary,25000.0
Department,It
Experience,2.0
City,Jaipur


In [52]:
df["Department"].value_counts()

,count
Department,
It,4
Hr,1


In [53]:
df[df['Department'] == "It"]

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,It,2.0,Jaipur
1,Virat Kohli,45000.0,It,5.0,Delhi
3,Rohit Sharma,25000.0,It,2.0,Mumbai
4,Shubman Gill,60000.0,It,3.0,Delhi


In [54]:
df

,Employee_Name,Salary,Department,Experience,City
0,Rohit Sharma,25000.0,It,2.0,Jaipur
1,Virat Kohli,45000.0,It,5.0,Delhi
2,Rahul,38750.0,Hr,3.0,Jaipur
3,Rohit Sharma,25000.0,It,2.0,Mumbai
4,Shubman Gill,60000.0,It,3.0,Delhi


In [56]:
df.isna().sum()

,0
Employee_Name,0
Salary,0
Department,0
Experience,0
City,0


## Key Insights

- The dataset contains 5 employees.
- The average salary is ₹38,750.
- The highest salary is ₹60,000.
- The average employee experience is 3 years.
- IT is the largest department with 4 employees.
- Jaipur and Delhi have the highest employee count, with 2 employees each.